torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -400,8 +401,17 @@ void Reducer::mark_variable_ready(VariableIndex index) {
     }
     local_used_work_ = process_group_->allreduce(local_used_maps_dev_);
 
+    // The autograd engine uses the default stream when running callbacks, so we
+    // pass in the current CUDA stream in case it is not the default.
+    c10::DeviceType deviceType = replica.contents.device().type();
+    const c10::impl::VirtualGuardImpl guard =
+        c10::impl::VirtualGuardImpl{deviceType};
+    const c10::Stream currentStream =
+        guard.getStream(replica.contents.device());
```

Comment:
Does this still work if this is CPU only DDP? Or is it because we need to support DDP so that we cannot directly call `getCurrentCUDAStream` from `CUDAStream`?

Reply:
Yes, we could've done this with `at::cuda::getCurrentCudaStream` and then set that as the guard below, although this wouldn't work for the CPU use case.

This approach here works for CPU or GPU, as the `VirtualGuardImpl` simply dispatches the call to the relevant guard implementation `CPUGuardImpl` for CPU and `CUDAGuardImpl` for CUDA. The former is effectively a no-op since there isn't a concept of streams in CPU.

CPU DDP should work fine with this, and I verified by running the DDP CPU test as well.